TestTemplateLevelHooks functional test is flaky

Flaky functional test. Example: https://github.com/argoproj/argo-workflows/actions/runs/3053610431/jobs/4924530479

```

=== RUN   TestHooksSuite/TestTemplateLevelHooks
Submitting workflow  lifecycle-hook-tmpl-level-
Waiting 1m0s for workflow metadata.name=lifecycle-hook-tmpl-level-79rph
 ? lifecycle-hook-tmpl-level-79rph Workflow 0s      

 ● lifecycle-hook-tmpl-level-79rph   Workflow  0s      
 └ ● lifecycle-hook-tmpl-level-79rph Steps     0s      
 └ ◷ step-1                          HTTP      0s      
 └ ● [0]                             StepGroup 0s      

 ● lifecycle-hook-tmpl-level-79rph   Workflow  0s      
 └ ● lifecycle-hook-tmpl-level-79rph Steps     0s      
 └ ✖ step-1                          HTTP      4s      Get "http://httpstat.us/": dial tcp: lookup httpstat.us on 10.43.0.10:53: no such host
 └ ● [0]                             StepGroup 0s      

 ● lifecycle-hook-tmpl-level-79rph   Workflow  0s      
 └ ● lifecycle-hook-tmpl-level-79rph Steps     0s      
 └ ✖ step-1                          HTTP      5s      Get "http://httpstat.us/": dial tcp: lookup httpstat.us on 10.43.0.10:53: no such host
 └ ● [0]                             StepGroup 0s      
 └ ◷ step-1.onExit                   HTTP      0s      

 ● lifecycle-hook-tmpl-level-79rph   Workflow  0s      
 └ ● lifecycle-hook-tmpl-level-79rph Steps     0s      
 └ ✖ step-1                          HTTP      6s      Get "http://httpstat.us/": dial tcp: lookup httpstat.us on 10.43.0.10:53: no such host
 └ ● [0]                             StepGroup 0s      
 └ ✖ step-1.onExit                   HTTP      1s      Get "http://httpstat.us/": dial tcp: lookup httpstat.us on 10.43.0.10:53: server misbehaving

 ✖ lifecycle-hook-tmpl-level-79rph   Workflow  7s      child 'lifecycle-hook-tmpl-level-79rph-3555890967' failed
 └ ✖ lifecycle-hook-tmpl-level-79rph Steps     7s      child 'lifecycle-hook-tmpl-level-79rph-3555890967' failed
 └ ✖ step-1                          HTTP      7s      Get "http://httpstat.us/": dial tcp: lookup httpstat.us on 10.43.0.10:53: no such host
 └ ✖ [0]                             StepGroup 7s      child 'lifecycle-hook-tmpl-level-79rph-3555890967' failed
 └ ✖ step-1.onExit                   HTTP      2s      Get "http://httpstat.us/": dial tcp: lookup httpstat.us on 10.43.0.10:53: server misbehaving

    hooks_test.go:83: condition never and cannot be met because the workflow is done
Checking expectation lifecycle-hook-tmpl-level-79rph
lifecycle-hook-tmpl-level-79rph : Failed child 'lifecycle-hook-tmpl-level-79rph-3555890967' failed
    hooks_test.go:86: 
        	Error Trace:	/home/runner/work/argo-workflows/argo-workflows/test/e2e/hooks_test.go:86
        	            				/home/runner/work/argo-workflows/argo-workflows/test/e2e/then.go:68
        	            				/home/runner/work/argo-workflows/argo-workflows/test/e2e/then.go:43
        	            				/home/runner/work/argo-workflows/argo-workflows/test/e2e/hooks_test.go:85
        	Error:      	Not equal: 
        	            	expected: "Succeeded"
        	            	actual  : "Failed"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-(v1alpha1.WorkflowPhase) (len=9) "Succeeded"
        	            	+(v1alpha1.WorkflowPhase) (len=6) "Failed"
        	            	 
        	Test:       	TestHooksSuite/TestTemplateLevelHooks
Checking expectation lifecycle-hook-tmpl-level-79rph
lifecycle-hook-tmpl-level-79rph : Failed child 'lifecycle-hook-tmpl-level-79rph-3555890967' failed
Found node id=lifecycle-hook-tmpl-level-79rph-2205883422 type=HTTP
    hooks_test.go:90: 
        	Error Trace:	/home/runner/work/argo-workflows/argo-workflows/test/e2e/hooks_test.go:90
        	            				/home/runner/work/argo-workflows/argo-workflows/test/e2e/then.go:110
        	            				/home/runner/work/argo-workflows/argo-workflows/test/e2e/then.go:68
        	            				/home/runner/work/argo-workflows/argo-workflows/test/e2e/then.go:85
        	            				/home/runner/work/argo-workflows/argo-workflows/test/e2e/hooks_test.go:87
        	Error:      	Not equal: 
        	            	expected: "Succeeded"
        	            	actual  : "Failed"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,2 +1,2 @@
        	            	-(v1alpha1.NodePhase) (len=9) "Succeeded"
        	            	+(v1alpha1.NodePhase) (len=6) "Failed"
        	            	 
        	Test:       	TestHooksSuite/TestTemplateLevelHooks
=== FAIL: HooksSuite/TestTemplateLevelHooks
FAIL	github.com/argoproj/argo-workflows/v3/test/e2e	172.192s
FAIL
make: *** [Makefile:523: test-functional] Error 1
```

--- a/test/e2e/hooks_test.go
+++ b/test/e2e/hooks_test.go
@@ -8,8 +8,8 @@ import (
 	"testing"
 
 	"github.com/stretchr/testify/assert"
-	apiv1 "k8s.io/api/core/v1"
 	"github.com/stretchr/testify/suite"
+	apiv1 "k8s.io/api/core/v1"
 	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 	"github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
@@ -53,14 +53,14 @@ spec:
 		ExpectWorkflow(func(t *testing.T, metadata *v1.ObjectMeta, status *v1alpha1.WorkflowStatus) {
 			assert.Equal(t, status.Phase, v1alpha1.WorkflowSucceeded)
 		}).ExpectWorkflowNode(func(status v1alpha1.NodeStatus) bool {
-			return strings.Contains(status.Name, ".hooks.running")
-		}, func(t *testing.T, status *v1alpha1.NodeStatus, pod *apiv1.Pod) {
-			assert.Equal(t, v1alpha1.NodeSucceeded, status.Phase)
-		}).ExpectWorkflowNode(func(status v1alpha1.NodeStatus) bool {
-			return strings.Contains(status.Name, ".hooks.succeed")
-		}, func(t *testing.T, status *v1alpha1.NodeStatus, pod *apiv1.Pod) {
-			assert.Equal(t, v1alpha1.NodeSucceeded, status.Phase)
-		})
+		return strings.Contains(status.Name, ".hooks.running")
+	}, func(t *testing.T, status *v1alpha1.NodeStatus, pod *apiv1.Pod) {
+		assert.Equal(t, v1alpha1.NodeSucceeded, status.Phase)
+	}).ExpectWorkflowNode(func(status v1alpha1.NodeStatus) bool {
+		return strings.Contains(status.Name, ".hooks.succeed")
+	}, func(t *testing.T, status *v1alpha1.NodeStatus, pod *apiv1.Pod) {
+		assert.Equal(t, v1alpha1.NodeSucceeded, status.Phase)
+	})
 }
 
 func (s *HooksSuite) TestWorkflowLevelHooksFailVersion() {
@@ -101,14 +101,14 @@ spec:
 		ExpectWorkflow(func(t *testing.T, metadata *v1.ObjectMeta, status *v1alpha1.WorkflowStatus) {
 			assert.Equal(t, status.Phase, v1alpha1.WorkflowFailed)
 		}).ExpectWorkflowNode(func(status v1alpha1.NodeStatus) bool {
-			return strings.Contains(status.Name, ".hooks.running")
-		}, func(t *testing.T, status *v1alpha1.NodeStatus, pod *apiv1.Pod) {
-			assert.Equal(t, v1alpha1.NodeSucceeded, status.Phase)
-		}).ExpectWorkflowNode(func(status v1alpha1.NodeStatus) bool {
-			return strings.Contains(status.Name, ".hooks.failed")
-		}, func(t *testing.T, status *v1alpha1.NodeStatus, pod *apiv1.Pod) {
-			assert.Equal(t, v1alpha1.NodeSucceeded, status.Phase)
-		})
+		return strings.Contains(status.Name, ".hooks.running")
+	}, func(t *testing.T, status *v1alpha1.NodeStatus, pod *apiv1.Pod) {
+		assert.Equal(t, v1alpha1.NodeSucceeded, status.Phase)
+	}).ExpectWorkflowNode(func(status v1alpha1.NodeStatus) bool {
+		return strings.Contains(status.Name, ".hooks.failed")
+	}, func(t *testing.T, status *v1alpha1.NodeStatus, pod *apiv1.Pod) {
+		assert.Equal(t, v1alpha1.NodeSucceeded, status.Phase)
+	})
 }
 
 func TestHooksSuite(t *testing.T) {

--- a/test/e2e/hooks_test.go
+++ b/test/e2e/hooks_test.go
@@ -1,5 +1,5 @@
-//go:build functional || local
-// +build functional local
+//go:build functional
+// +build functional
 
 package e2e
 

--- a/test/e2e/hooks_test.go
+++ b/test/e2e/hooks_test.go
@@ -1,5 +1,5 @@
-//go:build functional
-// +build functional
+//go:build functional || local
+// +build functional local
 
 package e2e
 
@@ -9,6 +9,7 @@ import (
 
 	"github.com/stretchr/testify/assert"
 	apiv1 "k8s.io/api/core/v1"
+	"github.com/stretchr/testify/suite"
 	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 	"github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
@@ -19,7 +20,7 @@ type HooksSuite struct {
 	fixtures.E2ESuite
 }
 
-func (s *HooksSuite) TestWorkflowLevelHooks() {
+func (s *HooksSuite) TestWorkflowLevelHooksSuccessVersion() {
 	s.Given().
 		Workflow(`apiVersion: argoproj.io/v1alpha1
 kind: Workflow
@@ -28,20 +29,23 @@ metadata:
 spec:
   entrypoint: main
   hooks:
-    exit:
-      template: http
     running:
       expression: workflow.status == "Running"
-      template: http
+      template: argosay
+    succeed:
+      expression: workflow.status == "Succeeded"
+      template: argosay
+
   templates:
     - name: main
       steps:
       - - name: step1
-          template: http
+          template: argosay
 
-    - name: http
-      http:
-        url: "http://httpstat.us"
+    - name: argosay
+      container:
+        image: argoproj/argosay:v2
+        command: ["/argosay"]
 `).When().
 		SubmitWorkflow().
 		WaitForWorkflow(fixtures.ToBeSucceeded).
@@ -49,48 +53,64 @@ spec:
 		ExpectWorkflow(func(t *testing.T, metadata *v1.ObjectMeta, status *v1alpha1.WorkflowStatus) {
 			assert.Equal(t, status.Phase, v1alpha1.WorkflowSucceeded)
 		}).ExpectWorkflowNode(func(status v1alpha1.NodeStatus) bool {
-		return strings.Contains(status.Name, "hook")
-	}, func(t *testing.T, status *v1alpha1.NodeStatus, pod *apiv1.Pod) {
-		assert.Equal(t, v1alpha1.NodeSucceeded, status.Phase)
-	})
+			return strings.Contains(status.Name, ".hooks.running")
+		}, func(t *testing.T, status *v1alpha1.NodeStatus, pod *apiv1.Pod) {
+			assert.Equal(t, v1alpha1.NodeSucceeded, status.Phase)
+		}).ExpectWorkflowNode(func(status v1alpha1.NodeStatus) bool {
+			return strings.Contains(status.Name, ".hooks.succeed")
+		}, func(t *testing.T, status *v1alpha1.NodeStatus, pod *apiv1.Pod) {
+			assert.Equal(t, v1alpha1.NodeSucceeded, status.Phase)
+		})
 }
 
-func (s *HooksSuite) TestTemplateLevelHooks() {
+func (s *HooksSuite) TestWorkflowLevelHooksFailVersion() {
 	s.Given().
 		Workflow(`apiVersion: argoproj.io/v1alpha1
 kind: Workflow
 metadata:
-  generateName: lifecycle-hook-tmpl-level-
+  generateName: lifecycle-hook-
 spec:
   entrypoint: main
+  hooks:
+    running:
+      expression: workflow.status == "Running"
+      template: hook
+    failed:
+      expression: workflow.status == "Failed"
+      template: hook
+
   templates:
     - name: main
       steps:
-        - - name: step-1
-            hooks:
-              exit:
-                template: http
-              success:
-                expression: steps["step-1"].status == "Succeeded"
-                template: http
-            template: http
-    - name: http
-      http:
-        url: "http://httpstat.us"
+      - - name: step1
+          template: argosay
+
+    - name: argosay
+      container:
+        image: argoproj/argosay:v2
+        command: ["/argosay; exit 1"]
+
+    - name: hook
+      container:
+        image: argoproj/argosay:v2
+        command: ["/argosay"]
 `).When().
 		SubmitWorkflow().
-		WaitForWorkflow(fixtures.ToBeSucceeded).
+		WaitForWorkflow(fixtures.ToBeFailed).
 		Then().
 		ExpectWorkflow(func(t *testing.T, metadata *v1.ObjectMeta, status *v1alpha1.WorkflowStatus) {
-			assert.Equal(t, v1alpha1.WorkflowSucceeded, status.Phase)
+			assert.Equal(t, status.Phase, v1alpha1.WorkflowFailed)
 		}).ExpectWorkflowNode(func(status v1alpha1.NodeStatus) bool {
-		return strings.Contains(status.Name, "hook")
-	}, func(t *testing.T, status *v1alpha1.NodeStatus, pod *apiv1.Pod) {
-		assert.Equal(t, v1alpha1.NodeSucceeded, status.Phase)
-	})
+			return strings.Contains(status.Name, ".hooks.running")
+		}, func(t *testing.T, status *v1alpha1.NodeStatus, pod *apiv1.Pod) {
+			assert.Equal(t, v1alpha1.NodeSucceeded, status.Phase)
+		}).ExpectWorkflowNode(func(status v1alpha1.NodeStatus) bool {
+			return strings.Contains(status.Name, ".hooks.failed")
+		}, func(t *testing.T, status *v1alpha1.NodeStatus, pod *apiv1.Pod) {
+			assert.Equal(t, v1alpha1.NodeSucceeded, status.Phase)
+		})
 }
 
 func TestHooksSuite(t *testing.T) {
-	// TODO: Tests are temporarily disabled: "https://github.com/argoproj/argo-workflows/issues/9591"
-	//suite.Run(t, new(HooksSuite))
+	suite.Run(t, new(HooksSuite))
 }

--- a/workflow/controller/hooks.go
+++ b/workflow/controller/hooks.go
@@ -11,7 +11,8 @@ import (
 	"github.com/argoproj/argo-workflows/v3/workflow/templateresolution"
 )
 
-func (woc *wfOperationCtx) executeWfLifeCycleHook(ctx context.Context, tmplCtx *templateresolution.Context) error {
+func (woc *wfOperationCtx) executeWfLifeCycleHook(ctx context.Context, tmplCtx *templateresolution.Context) (bool, error) {
+	var hookNodes []*wfv1.NodeStatus
 	for hookName, hook := range woc.execWf.Spec.Hooks {
 		//exit hook will be executed in runOnExitNode
 		if hookName == wfv1.ExitLifecycleEvent {
@@ -19,24 +20,30 @@ func (woc *wfOperationCtx) executeWfLifeCycleHook(ctx context.Context, tmplCtx *
 		}
 		execute, err := argoexpr.EvalBool(hook.Expression, env.GetFuncMap(template.EnvMap(woc.globalParams)))
 		if err != nil {
-			return err
+			return false, err
 		}
 		if execute {
 			hookNodeName := generateLifeHookNodeName(woc.wf.ObjectMeta.Name, string(hookName))
 			woc.log.WithField("lifeCycleHook", hookName).WithField("node", hookNodeName).Infof("Running workflow level hooks")
 			hookNode, err := woc.executeTemplate(ctx, hookNodeName, &wfv1.WorkflowStep{Template: hook.Template, TemplateRef: hook.TemplateRef}, tmplCtx, hook.Arguments, &executeTemplateOpts{})
 			if err != nil {
-				return err
+				return false, err
 			}
 			woc.addChildNode(woc.wf.Name, hookNodeName)
+			hookNodes = append(hookNodes, hookNode)
 			// If the hookNode node is HTTP template, it requires HTTP reconciliation, do it here
 			if hookNode != nil && woc.nodeRequiresTaskSetReconciliation(hookNode.Name) {
 				woc.taskSetReconciliation(ctx)
 			}
 		}
 	}
+	for _, hookNode := range hookNodes {
+		if !hookNode.Fulfilled() {
+			return false, nil
+		}
+	}
 
-	return nil
+	return true, nil
 }
 
 func (woc *wfOperationCtx) executeTmplLifeCycleHook(ctx context.Context, scope *wfScope, lifeCycleHooks wfv1.LifecycleHooks, parentNode *wfv1.NodeStatus, boundaryID string, tmplCtx *templateresolution.Context, prefix string) (bool, error) {

--- a/workflow/controller/hooks.go
+++ b/workflow/controller/hooks.go
@@ -20,14 +20,14 @@ func (woc *wfOperationCtx) executeWfLifeCycleHook(ctx context.Context, tmplCtx *
 		}
 		execute, err := argoexpr.EvalBool(hook.Expression, env.GetFuncMap(template.EnvMap(woc.globalParams)))
 		if err != nil {
-			return false, err
+			return true, err
 		}
 		if execute {
 			hookNodeName := generateLifeHookNodeName(woc.wf.ObjectMeta.Name, string(hookName))
 			woc.log.WithField("lifeCycleHook", hookName).WithField("node", hookNodeName).Infof("Running workflow level hooks")
 			hookNode, err := woc.executeTemplate(ctx, hookNodeName, &wfv1.WorkflowStep{Template: hook.Template, TemplateRef: hook.TemplateRef}, tmplCtx, hook.Arguments, &executeTemplateOpts{})
 			if err != nil {
-				return false, err
+				return true, err
 			}
 			woc.addChildNode(woc.wf.Name, hookNodeName)
 			hookNodes = append(hookNodes, hookNode)

--- a/workflow/controller/operator.go
+++ b/workflow/controller/operator.go
@@ -400,8 +400,6 @@ func (woc *wfOperationCtx) operate(ctx context.Context) {
 	// This strconv.Quote is necessary so that the escaped quotes are not removed during parameter substitution
 	woc.globalParams[common.GlobalVarWorkflowFailures] = strconv.Quote(string(failedNodeBytes))
 
-	woc.log.Errorf("@#@#@#@#@##@#@#@# Node Inspect Lifecycle 실행 이전 : %+v", node)
-
 	hookCompleted, err := woc.executeWfLifeCycleHook(ctx, tmplCtx)
 	if err != nil {
 		woc.markNodeError(node.Name, err)
@@ -413,9 +411,6 @@ func (woc *wfOperationCtx) operate(ctx context.Context) {
 	if !hookCompleted {
 		return
 	}
-
-	woc.log.Errorf("@#@#@#@#@##@#@#@# Node Inspect Lifecycle 실행 이후 : %+v", node)
-	woc.log.Errorf("############################################################")
 
 	if !node.Fulfilled() {
 		// node can be nil if a workflow created immediately in a parallelism == 0 state

--- a/workflow/controller/operator.go
+++ b/workflow/controller/operator.go
@@ -400,12 +400,22 @@ func (woc *wfOperationCtx) operate(ctx context.Context) {
 	// This strconv.Quote is necessary so that the escaped quotes are not removed during parameter substitution
 	woc.globalParams[common.GlobalVarWorkflowFailures] = strconv.Quote(string(failedNodeBytes))
 
-	err = woc.executeWfLifeCycleHook(ctx, tmplCtx)
+	woc.log.Errorf("@#@#@#@#@##@#@#@# Node Inspect Lifecycle 실행 이전 : %+v", node)
+
+	hookCompleted, err := woc.executeWfLifeCycleHook(ctx, tmplCtx)
 	if err != nil {
 		woc.markNodeError(node.Name, err)
 	}
 	// Reconcile TaskSet and Agent for HTTP templates
 	woc.taskSetReconciliation(ctx)
+
+	// Check all hooks are completes
+	if !hookCompleted {
+		return
+	}
+
+	woc.log.Errorf("@#@#@#@#@##@#@#@# Node Inspect Lifecycle 실행 이후 : %+v", node)
+	woc.log.Errorf("############################################################")
 
 	if !node.Fulfilled() {
 		// node can be nil if a workflow created immediately in a parallelism == 0 state

